internal/logic/logic.go

```diff
@@ -1,34 +1,201 @@
 	HealthCheck() bool
-	//Signup(model.SignUpCredentials) *respModel.Response
-	//Login(http.ResponseWriter, model.LoginCredentials) *respModel.Response
+	Signup(model.SignUpCredentials) *respModel.Response
+	Login(http.ResponseWriter, model.LoginCredentials) *respModel.Response
 	//UserData(any) *respModel.Response
 	//Activate(id any) *respModel.Response
 }
 
 type userMgmtSvcLogic struct {
-	DsSvc datasource.DataSourceI
-	//loginService jwtSvc.LoginService
-	//jwtService   jwtSvc.JWTService
+	DsSvc      datasource.DataSourceI
+	jwtService jwtSvc.JWTService
+	msgQueue   config.MsgQueue
+	cookie     config.CookieStruct
 }
 
-func NewUserMgmtSvcLogic(ds datasource.DataSourceI) UserMgmtSvcLogicIer {
+func NewUserMgmtSvcLogic(ds datasource.DataSourceI, jwtService jwtSvc.JWTService, msgQueue config.MsgQueue, cookie config.CookieStruct) UserMgmtSvcLogicIer {
 	return &userMgmtSvcLogic{
-		DsSvc: ds,
-		//loginService: loginSvc,
-		//jwtService:   jwtService,
+		DsSvc:      ds,
+		jwtService: jwtService,
+		msgQueue:   msgQueue,
+		cookie:     cookie,
 	}
 }
 
 func (l userMgmtSvcLogic) HealthCheck() bool {
 	// check all internal services are working fine
 	return l.DsSvc.HealthCheck()
 }
+
+func (l userMgmtSvcLogic) Signup(credential model.SignUpCredentials) *respModel.Response {
+	result, err := l.DsSvc.Get(map[string]interface{}{"email": credential.Email})
+	if err != nil {
+		log.Error(err.Error())
+		return &respModel.Response{
+			Status:  http.StatusInternalServerError,
+			Message: codes.GetErr(codes.ErrCreatingAccount),
+			Data:    nil,
+		}
+	}
+	if len(result) != 0 {
+		log.Error(codes.GetErr(codes.ErrEmailExists))
+		return &respModel.Response{
+			Status:  http.StatusBadRequest,
+			Message: codes.GetErr(codes.ErrEmailExists),
+			Data:    nil,
+		}
+	}
+	newUser := model.User{
+		Id:           uuid.New().String(),
+		Email:        credential.Email,
+		Name:         credential.Name,
+		Company:      "perennial",
+		RegisteredOn: credential.RegistrationTimestamp,
+	}
+
+	hashedPassword, err := crypto.GeneratePasswordHash([]byte(credential.Password), []byte(newUser.Id))
+	if err != nil {
+		log.Error(err.Error())
+		return &respModel.Response{
+			Status:  http.StatusInternalServerError,
+			Message: codes.GetErr(codes.ErrHashPassword),
+			Data:    nil,
+		}
+	}
+	newUser.Password = hashedPassword
+
+	err = l.DsSvc.Insert(newUser)
+	if err != nil {
+		log.Error(err.Error())
+		return &respModel.Response{
+			Status:  http.StatusInternalServerError,
+			Message: codes.GetErr(codes.ErrCreatingAccount),
+			Data:    nil,
+		}
+	}
+	go func() {
+		userID := fmt.Sprintf(`{"user_id":"%s"}`, newUser.Id)
+		err = l.msgQueue.MsgBroker.PushMsg(userID, l.msgQueue.PubId, l.msgQueue.Channel)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+		return
+	}()
+	return &respModel.Response{
+		Status:  http.StatusCreated,
+		Message: codes.GetErr(codes.Success),
+		Data:    codes.GetErr(codes.AccActivationInProcess),
```

Comment:
This is not an error. Just return the text as "Activation in progress" in the data field.

Reply:
done